app_test.go

```diff
@@ -473,3 +491,77 @@ func TestNopLogger(t *testing.T) {
 	app := fxtest.New(t, NopLogger)
 	app.RequireStart().RequireStop()
 }
+
+type testWrappedErr struct {
+	graph string
+}
+
+func (we testWrappedErr) Graph() string {
+	return we.graph
+}
+
+func (we testWrappedErr) Error() string {
+	return "great sadness"
+}
+
+func TestVisualizeError(t *testing.T) {
+	t.Run("NotWrappedError", func(t *testing.T) {
+		graph, err := VisualizeError(errors.New("great sadness"))
+		assert.Equal(t, "", graph)
+		require.Error(t, err)
+	})
+
+	t.Run("WrappedErrorWithEmptyGraph", func(t *testing.T) {
+		graph, err := VisualizeError(testWrappedErr{graph: ""})
+		assert.Equal(t, "", graph)
+		require.Error(t, err)
+	})
+
+	t.Run("WrappedError", func(t *testing.T) {
+		graph, err := VisualizeError(testWrappedErr{graph: "graph"})
+		assert.Equal(t, "graph", graph)
+		require.NoError(t, err)
+	})
+}
+
+type graphErrHandler struct {
+	errStr   string
+	graphStr string
+}
+
+func (eh *graphErrHandler) HandleError(err error) {
+	eh.errStr = err.Error()
+	eh.graphStr, _ = VisualizeError(err)
+}
+
+func TestErrorHook(t *testing.T) {
+	t.Run("UnvisualizableError", func(t *testing.T) {
+		type A struct{}
+
+		errHandler := graphErrHandler{}
+		fxtest.New(t,
+			Provide(func() A { return A{} }),
+			Invoke(func(A) error { return fmt.Errorf("great sadness") }),
+			ErrorHook(&errHandler),
+		)
+		assert.Equal(t, "great sadness", errHandler.errStr)
+		assert.Equal(t, "", errHandler.graphStr)
+
+	})
+
+	t.Run("GraphWithError", func(t *testing.T) {
+		type A struct{}
+		type B struct{}
+		type C struct{}
+
+		errHandler := graphErrHandler{}
+		fxtest.New(t,
+			Provide(func() (B, error) { return B{}, fmt.Errorf("great sadness") }),
+			Provide(func(B) A { return A{} }),
+			Invoke(func(A) {}),
+			ErrorHook(&errHandler),
+		)
+		assert.Contains(t, errHandler.graphStr, "\"fx_test.B\" [color=red];")
+		assert.Contains(t, errHandler.graphStr, "\"fx_test.A\" [color=orange];")
```

Comment:
instead of escaping double quotes, you can use backticks to let you write double quotes as-is

    assert.Contains(t, ..., `"fx_test.B" [color=red];`)

Reply:
That makes it a lot more readable 😄